solr/solrj/src/java/org/apache/solr/common/cloud/SolrZkClient.java

```diff
@@ -147,125 +154,93 @@ public SolrZkClient(
+      chroot = zkServerAddress.substring(chrootIndex + 1);
     }
-    this.zkClientConnectionStrategy = strat;
+    this.higherLevelIsClosed = higherLevelIsClosed;
 
-    if (!strat.hasZkCredentialsToAddAutomatically()) {
-      ZkCredentialsProvider zkCredentialsToAddAutomatically =
-          createZkCredentialsToAddAutomatically();
-      strat.setZkCredentialsToAddAutomatically(zkCredentialsToAddAutomatically);
+    if (zkCredentialsProvider == null) {
+      zkCredentialsProvider = createZkCredentialsToAddAutomatically();
+    }
+    if (aclProvider == null) {
+      aclProvider = createACLProvider();
+    }
+    if (chroot != null && aclProvider instanceof SecurityAwareZkACLProvider) {
+      this.aclProvider = ((SecurityAwareZkACLProvider)aclProvider).withChroot(chroot);
+    } else {
+      this.aclProvider = aclProvider;
     }
 
     this.zkClientTimeout = zkClientTimeout;
-    // we must retry at least as long as the session timeout
-    zkCmdExecutor =
-        new ZkCmdExecutor(
-            zkClientTimeout,
-            new IsClosed() {
-
-              @Override
-              public boolean isClosed() {
-                return SolrZkClient.this.isClosed();
-              }
-            });
-    connManager =
-        new ConnectionManager(
-            "ZooKeeperConnection Watcher:" + zkServerAddress,
-            this,
-            zkServerAddress,
-            strat,
-            onReconnect,
-            beforeReconnect,
-            new IsClosed() {
-
-              @Override
-              public boolean isClosed() {
-                return SolrZkClient.this.isClosed();
-              }
-            });
 
-    try {
-      strat.connect(
-          zkServerAddress,
-          zkClientTimeout,
-          wrapWatcher(connManager),
-          zooKeeper -> {
-            SolrZooKeeper oldKeeper = keeper;
-            keeper = zooKeeper;
-            try {
-              closeKeeper(oldKeeper);
-            } finally {
-              if (isClosed) {
-                // we may have been closed
-                closeKeeper(SolrZkClient.this.keeper);
-              }
-            }
-          });
-    } catch (Exception e) {
-      connManager.close();
-      if (keeper != null) {
-        try {
-          keeper.close();
-        } catch (InterruptedException e1) {
-          Thread.currentThread().interrupt();
-        }
-      }
-      throw new SolrException(SolrException.ErrorCode.SERVER_ERROR, e);
-    }
+    RetryPolicy retryPolicy = new ExponentialBackoffRetry(1000, 3);
+    var clientBuilder = CuratorFrameworkFactory.builder()
+        .ensembleProvider(new FixedEnsembleProvider(zkHost))
+        .namespace(chroot)
+        .sessionTimeoutMs(zkClientTimeout)
+        .connectionTimeoutMs(clientConnectTimeout)
+        .aclProvider(this.aclProvider)
+        .authorization(zkCredentialsProvider.getCredentials())
+        .retryPolicy(retryPolicy);
 
+    client = clientBuilder.build();
```

Comment:
I recently worked on an experimental hack for some days on 8x to have the Overseer use Curator for elections.  One thing that I got stuck on was that Solr's tests were finding some threads lingering.  Eventually I discovered that Curator has an internal executor and a way to set it on the builder `runSafeService`.  Once I did that and tended to ensuring this executor got shot down, I didn't have that problem anymore.  Just sharing this with you now in hopes it helps.

Reply:
FWIW https://github.com/apache/solr/pull/1743 used this runSafeService stuff for the hadoop-auth changes. I think it might only come into play when there is an external zkClient provided but good to know either way.